cmds/free/free.go

```diff
@@ -0,0 +1,241 @@
+// rounded, so the values are guaranteed to be "at least X"
+func humanReadableValue(value uint64) string {
+	v := value
+	// bits to shift. 0 means bytes, 10 means kB, and so on. 40 is the highest
+	// and it means tB
+	var shift uint
+	for {
+		if shift >= uint(len(units)*10) {
+			// 4 means tebibyte, we don't go further
+			break
+		}
+		if v/1024 < 1 {
+			break
+		}
+		v /= 1024
+		shift += 10
+	}
+	var decimal uint64
+	if shift > 0 {
+		// no rounding. Is there a better way to do this?
+		decimal = ((value - (value >> shift << shift)) >> (shift - 10)) * 1000 / 1024 / 100
+	}
+	return fmt.Sprintf("%v.%v%v",
+		value>>shift,
+		decimal,
+		units[shift/10],
+	)
+}
+
+// formatValueByConfig formats a size in bytes in the appropriate unit,
+// depending on whether FreeConfig specifies a human-readable format or a
+// specific unit
+func formatValueByConfig(value uint64, config *FreeConfig) string {
+	if config.HumanOutput {
+		return humanReadableValue(value)
+	}
+	// units and decimal part are not printed when a unit is explicitly specified
+	return fmt.Sprintf("%v", value>>config.Unit)
+}
+
+// Free prints physical memory and swap space information. The fields will be
+// expressed with the specified unit (e.g. KB, MB)
+func Free(config *FreeConfig) error {
+	m, err := meminfo()
+	if err != nil {
+		log.Fatal(err)
+	}
+	mmi, err := getMainMemInfo(m, config)
+	if err != nil {
+		return err
+	}
+	si, err := getSwapInfo(m, config)
+	if err != nil {
+		return err
+	}
+	mi := MemInfo{Mem: *mmi, Swap: *si}
+	if config.ToJSON {
+		jsonData, err := json.Marshal(mi)
+		if err != nil {
+			return err
+		}
+		fmt.Println(string(jsonData))
+	} else {
+		fmt.Printf("              total        used        free      shared  buff/cache   available\n")
+		fmt.Printf("%-7s %11v %11v %11v %11v %11v %11v\n",
+			"Mem:",
+			formatValueByConfig(mmi.Total, config),
+			formatValueByConfig(mmi.Used, config),
+			formatValueByConfig(mmi.Free, config),
+			formatValueByConfig(mmi.Shared, config),
+			formatValueByConfig(mmi.Buffers+mmi.Cached, config),
+			formatValueByConfig(mmi.Available, config),
+		)
+		fmt.Printf("%-7s %11v %11v %11v\n",
+			"Swap:",
+			formatValueByConfig(si.Total, config),
+			formatValueByConfig(si.Used, config),
+			formatValueByConfig(si.Free, config),
+		)
+	}
+	return nil
+}
+
+// validateUnits checks that only one option of -b, -k, -m, -g, -t or -h has been
+// specified on the command line
+func validateUnits() bool {
+	count := 0
+	if *inBytes {
+		count++
+	}
+	if *inKB {
+		count++
+	}
+	if *inMB {
+		count++
+	}
+	if *inGB {
+		count++
+	}
+	if *inTB {
+		count++
+	}
+	if *humanOutput {
+		count++
+	}
+	if count > 1 {
+		return false
+	}
+	return true
+}
+
+func main() {
+	flag.Parse()
+	if !validateUnits() {
+		log.Fatal("Options -k, -m, -g, -t and -h are mutually exclusive")
+	}
+	config := FreeConfig{ToJSON: *toJSON}
+	if *humanOutput {
+		config.HumanOutput = true
+	} else {
+		var unit unit = KB
```

Comment:
super nit, and maybe a go feature you haven't seen yet:

```go
switch {
case *inBytes:
  unit = B
case *inKB:
  unit = KB
case *inMB:
  unit = MB
}
```

Reply:
oh. I know `switch` but I didn't think I could use it that way. Coming in a separate PR with the var grouping above